### PR TITLE
[ci] fix some shellcheck warnings in package-building scripts

### PR DESCRIPTION
**Pull request by vnherdeiro**
Contributes to https://github.com/microsoft/LightGBM/issues/6498 continues from #6621

Now I am tackling `./build-python.sh` and `./build-cran-package.sh`

All feedback is welcome!

Please forgive the polluted commit history.

### CODE REVIEW

File: build-cran-package.sh [hunk 1]
```diff
@@ -91,10 +91,10 @@ mkdir -p "${EIGEN_R_DIR}"
 
 modules="Cholesky Core Dense Eigenvalues Geometry Householder Jacobi LU QR SVD"
 for eigen_module in ${modules}; do
-    cp external_libs/eigen/Eigen/${eigen_module} "${EIGEN_R_DIR}/${eigen_module}"
-    if [ ${eigen_module} != "Dense" ]; then
+    cp external_libs/eigen/Eigen/"${eigen_module}" "${EIGEN_R_DIR}/${eigen_module}"
```
Comment (jameslamb) [inline build-cran-package.sh:94]:
```suggestion
    cp "external_libs/eigen/Eigen/${eigen_module}" "${EIGEN_R_DIR}/${eigen_module}"
```

Can you please wrap entire paths in `"` instead of inserting the `"` in the middle of them? I think that's a little easier to read (personal preference).

File: build-cran-package.sh [hunk 2]
```diff
@@ -91,10 +91,10 @@ mkdir -p "${EIGEN_R_DIR}"
...
+    if [ "${eigen_module}" != "Dense" ]; then
         mkdir -p "${EIGEN_R_DIR}/src/${eigen_module}/"
-        cp -R external_libs/eigen/Eigen/src/${eigen_module}/* "${EIGEN_R_DIR}/src/${eigen_module}/"
+        cp -R external_libs/eigen/Eigen/src/"${eigen_module}"/* "${EIGEN_R_DIR}/src/${eigen_module}/"
```
Comment (jameslamb) [inline build-cran-package.sh:97]:
```suggestion
        cp -R "external_libs/eigen/Eigen/src/${eigen_module}"/* "${EIGEN_R_DIR}/src/${eigen_module}/"
```

File: build-cran-package.sh [hunk 3]
```diff
@@ -130,7 +130,7 @@ cd "${TEMP_R_DIR}"
     # Rtools35 (used with R 3.6 on Windows) doesn't support C++17
     LGB_CXX_STD="C++17"
     using_windows_and_r3=$(
-        Rscript -e 'cat(.Platform$OS.type == "windows" && R.version[["major"]] < 4)'
+        Rscript -e "cat(.Platform$OS.type == \"windows\" && R.version[[\"major\"]] < 4)"
```
Comment (jameslamb) [inline build-cran-package.sh:133]: Did `shellcheck` suggest this? What was the warning/error code and text?

If yes, could you please just ignore that check for this line with a comment similar to https://github.com/microsoft/LightGBM/blob/41ba9e8f00c89d72e5cb71c964722ce1ed4d8445/docs/build-docs.sh#L23

This is intentionally using single quotes so that inner `$` is treated as a literal `$` and not used for string interpolation by the shell. I strongly suspect this change is going to break this line of code otherwise.

File: build-cran-package.sh [hunk 4]
```diff
@@ -151,9 +151,8 @@ cd "${TEMP_R_DIR}"
         -e 's/^.*#pragma region.*$//' \
         -e 's/^.*#pragma endregion.*$//' \
         -e 's/^.*#pragma warning.*$//' \
-        "${file}"
-    done
-    find . -name '*.h.bak' -o -name '*.hpp.bak' -o -name '*.cpp.bak' -exec rm {} \;
+        {} +
+    find . \( -name '*.h.bak' -o -name '*.hpp.bak' -o -name '*.cpp.bak' \) -exec rm {} \;
```
Comment (jameslamb) [inline build-cran-package.sh:155]:
```suggestion
    find . -name '*.bak' -exec rm {} \;
```

While you're touching this anyway, let's just change this to deleting all `.bak` files from this directory and below. None of them are necessary, they're all just temporary side effects of the `sed` usage.

Comment (jameslamb) [reply build-cran-package.sh:155]: and actually.... could you move this down to where it says "Cleaning sed backup files", replacing this `rm`?

https://github.com/microsoft/LightGBM/blob/41ba9e8f00c89d72e5cb71c964722ce1ed4d8445/build-cran-package.sh#L167

Might as well just do that only 1 time and catch all of them.

File: build-python.sh [hunk 5]
```diff
@@ -1,4 +1,4 @@
-#!/bin/sh
+#!/bin/bash
```
Comment (jameslamb) [inline build-python.sh:1]: please revert this. This script is very intentionally using `sh`, so that it's as portable as possible (`bash` on Windows has some eccentricities we'd prefer to avoid).

I'll wait to review the rest of the changes on this file until after you've done that, in case some of them were just a result of switching to bash.

**Review by jameslamb — CHANGES_REQUESTED** [review]
Thanks for working on this!

Left a first round of comments for your consideration.

**Comment by jameslamb** [comment]
> *Please forgive the polluted commit history.*

No worries! We'll squash all of the history when this is merged anyway.

**Comment by vnherdeiro** [comment]
That is a prompt review @jameslamb Thanks, will amend soonish!

File: build-cran-package.sh (hunk 3 above)
Comment (vnherdeiro) [reply build-cran-package.sh:133]: Yeah I see the change has broken some windows CIs, will revert

File: build-python.sh [hunk 6]
```diff
@@ -308,12 +308,16 @@ if test "${INSTALL}" = true; then
...
-        echo 'build-backend = "setuptools.build_meta"' >> ./pyproject.toml
-        echo "" >> ./pyproject.toml
-        echo "recursive-include lightgbm *.dll *.dylib *.so" > ./MANIFEST.in
-        echo "" >> ./MANIFEST.in
+        {
+            echo '[build-system]'; \
+            echo 'requires = ["setuptools"]'; \
+            echo 'build-backend = "setuptools.build_meta"'; \
+            echo ""; \
+        }  >> ./pyproject.toml
```
Comment (jameslamb) [inline build-python.sh:316]: what is the purpose of these `{}`? Can you please share the exact `shellcheck` warnings this is addressing?

I don't understand the purpose of doing this in a sub-shell, I strongly suspect there's a better way to address whatever warning this is about.

File: build-python.sh [hunk 7]
```diff
@@ -342,7 +346,7 @@ if test "${BUILD_SDIST}" = true; then
     python -m build \
         --sdist \
         --outdir ../dist \
-        ${BUILD_ARGS} \
+        ${BUILD_ARGS:+ -o "$BUILD_ARGS"} \
```
Comment (jameslamb) [inline build-python.sh:349]: I don't understand this change, can you please explain it?

File: build-python.sh [hunk 8]
```diff
@@ -352,26 +356,26 @@ if test "${BUILD_WHEEL}" = true; then
...
         .
 fi
 
 if test "${INSTALL}" = true; then
     echo "--- installing lightgbm ---"
     cd ../dist
     if test "${BUILD_WHEEL}" = true; then
-        PACKAGE_NAME="lightgbm*.whl"
+        PACKAGE_NAME=$(find . -name "lightgbm*.whl" -print -quit)
```
Comment (jameslamb) [inline build-python.sh:367]:
```suggestion
        PACKAGE_NAME="$(echo lightgbm*.whl)"
```

I think `echo` would be simpler here.

File: build-python.sh [hunk 9]
```diff
@@ -352,26 +356,26 @@ if test "${BUILD_WHEEL}" = true; then
...
     else
-        PACKAGE_NAME="lightgbm*.tar.gz"
+        PACKAGE_NAME=$(find . -name "lightgbm*.tar.gz" -print -quit)
```
Comment (jameslamb) [inline build-python.sh:369]:
```suggestion
        PACKAGE_NAME="$(echo lightgbm*.tar.gz)"
```

File: build-python.sh [hunk 10]
```diff
@@ -254,14 +254,14 @@ create_isolated_source_dir() {
     modules="Cholesky Core Dense Eigenvalues Geometry Householder Jacobi LU QR SVD"
     for eigen_module in ${modules}; do
         cp \
-            external_libs/eigen/Eigen/${eigen_module} \
-            ./lightgbm-python/external_libs/eigen/Eigen/${eigen_module}
-        if [ ${eigen_module} != "Dense" ]; then
-            mkdir -p ./lightgbm-python/external_libs/eigen/Eigen/src/${eigen_module}/
+            external_libs/eigen/Eigen/"${eigen_module}" \
+            ./lightgbm-python/external_libs/eigen/Eigen/"${eigen_module}"
```
Comment (jameslamb) [inline build-python.sh:258]:
```suggestion
            "external_libs/eigen/Eigen/${eigen_module}" \
            "./lightgbm-python/external_libs/eigen/Eigen/${eigen_module}"
```

Please, no `"` in the middle of filepaths.

File: build-python.sh [hunk 11]
```diff
@@ -254,14 +254,14 @@ create_isolated_source_dir() {
...
+        if [ "${eigen_module}" != "Dense" ]; then
+            mkdir -p ./lightgbm-python/external_libs/eigen/Eigen/src/"${eigen_module}"/
```
Comment (jameslamb) [inline build-python.sh:260]:
```suggestion
            mkdir -p "./lightgbm-python/external_libs/eigen/Eigen/src/${eigen_module}/"
```

File: build-python.sh [hunk 12]
```diff
@@ -254,14 +254,14 @@ create_isolated_source_dir() {
...
             cp \
                 -R \
-                external_libs/eigen/Eigen/src/${eigen_module}/* \
-                ./lightgbm-python/external_libs/eigen/Eigen/src/${eigen_module}/
+                external_libs/eigen/Eigen/src/"${eigen_module}"/* \
+                ./lightgbm-python/external_libs/eigen/Eigen/src/"${eigen_module}"/
```
Comment (jameslamb) [inline build-python.sh:264]:
```suggestion
                "external_libs/eigen/Eigen/src/${eigen_module}"/* \
                "./lightgbm-python/external_libs/eigen/Eigen/src/"${eigen_module}/"
```

**Review by jameslamb — CHANGES_REQUESTED** [review]
Thanks, I've left another round of questions and suggestions. I'm really confused by some of these changes... I left comments asking for clarification on a few.

File: build-python.sh (hunk 6 above)
Comment (vnherdeiro) [reply build-python.sh:316]: The subsequence commands piping into the same file where raising this warning https://www.shellcheck.net/wiki/SC2129

I did it with a subshell as suggested, joining all strings with `\n` and single echo should be fine too

**Comment by vnherdeiro** [comment]
@jameslamb thanks for the review and code suggestions about the more extent quotemarks! Mind if I start a new PR soon with changes for the files you had asked my to postpone from https://github.com/microsoft/LightGBM/pull/6621 :

- .ci/test.sh
- .ci/test-r-package.sh
- .ci/setup.sh

File: build-python.sh (hunk 6 above)
Comment (jameslamb) [reply build-python.sh:316]: Thanks for that explanation. I really find this command-group style difficult to read. 

As it says there

> *This is mainly a stylistic issue, and can freely be ignored.*

So please revert all of the changes like this and add `# shellcheck disable=SC2129` comments to suppress these.

File: build-python.sh [hunk 13]
```diff
@@ -352,26 +356,26 @@ if test "${BUILD_WHEEL}" = true; then
     python -m build \
         --wheel \
         --outdir ../dist \
-        ${BUILD_ARGS} \
+        ${BUILD_ARGS:+${BUILD_ARGS}} \
```
Comment (jameslamb) [inline build-python.sh:359]: Please, explain this. What specifically is `shellcheck` complaining about that led to this change?

File: build-python.sh [hunk 14]
```diff
@@ -352,26 +356,26 @@ if test "${BUILD_WHEEL}" = true; then
...
-        PACKAGE_NAME="lightgbm*.whl"
+        PACKAGE_NAME="$(echo lightgbm*.whl)"
     else
-        PACKAGE_NAME="lightgbm*.tar.gz"
+        PACKAGE_NAME="$(echo lightgbm*.tar.gz)"
     fi
     # ref for use of '--find-links': https://stackoverflow.com/a/52481267/3986677
     pip install \
-        ${PIP_INSTALL_ARGS} \
+        ${PIP_INSTALL_ARGS:+ -o "$PIP_INSTALL_ARGS"} \
```
Comment (jameslamb) [inline build-python.sh:373]: What is the purpose of this change?

File: build-python.sh [hunk 15]
```diff
@@ -91,42 +91,42 @@ while [ $# -gt 0 ]; do
     # customized library paths #
     ############################
     --boost-dir|--boost-dir=*)
-        if [[ "$1" != *=* ]];
+        if echo "$1" | grep -q '^*=*$';
```
Comment (jameslamb) [inline build-python.sh:94]: What is `shellcheck` complaining about here?

**Review by jameslamb — DISMISSED** [review]
Thanks for working on this.

I'm confused by several of these changes. As I've asked in previous reviews... please explain the reasoning for your proposes changes here. Several of these are not obvious at all, and it's difficult to review without knowing the underlying `shellcheck` warnings you're trying to address.

**Comment by jameslamb** [comment]
> *Do you mind if I start a new PR soon with changes for the files you had asked my to postpone from https://github.com/microsoft/LightGBM/pull/6621*

Looks like you decided to do that in #6646. That's fine, will review there.

File: build-python.sh (hunk 7 above)
Comment (jameslamb) [reply build-python.sh]: @vnherdeiro are you still interested in this work and available to help? If not, could you please let us know so we can continue it in other PRs?

**Comment by jameslamb** [comment]
@vnherdeiro will you have time in the next week to come back and work with us on this? If not, I'll close this and fix the remaining `shellcheck` issues myself.

**Comment by jameslamb** [comment]
Alright thanks. You can start here: https://github.com/microsoft/LightGBM/pull/6641#discussion_r1775446190

**Comment by vnherdeiro** [comment]
@jameslamb I am still figuring out how github review works, I had input comments and responses to your own but missed the submit review action, hence they were not visitble to you (or anyone but me). Completed that now.

File: build-python.sh (hunk 7 above)
Comment (jameslamb) [reply build-python.sh]: Thanks for that.

The `${BUILD_ARGS:+"${BUILD_ARGS}"}` form is not what we want for this script. This warning is about quoting to "prevent word splitting" ... but we genuinely **want** word splitting here, so that this string containing multiple command line arguments (separated by spaces) will be passed through to the underlying commands that way, not as a single long string to one positional argument.

And we cannot use arrays because this is `sh`, not `bash`.

I just pushed a commit reverting these changes and just adding `#shellcheck disable` comments.

File: build-python.sh (hunk 13 above)
Comment (jameslamb) [reply build-python.sh]: Thanks, but I just pushed commits reverting this for the reasons mentioned in https://github.com/microsoft/LightGBM/pull/6641#discussion_r1804007090.

The claim you've made here that this being empty would result in the command failing is simply not true. You can try this yourself:

```shell
sh build-python.sh bdist_wheel
```

That will reach this line with `BUILD_ARGS` empty, and nothing will fail. We even intentionally set `BUILD_ARGS` to the empty string when `--precompile` is passed.

https://github.com/microsoft/LightGBM/blob/668bf5dadf1eb9a846302b2b76a313fbbef52870/build-python.sh#L295-L299

And I use `sh build-python.sh install --precompile` every time I develop on LightGBM's Python package.

File: build-python.sh (hunk 14 above)
Comment (jameslamb) [reply build-python.sh]: > I had explained it here-above

I see. I had not seen https://github.com/microsoft/LightGBM/pull/6641#discussion_r1748393161 when I wrote this, maybe it's another one that wasn't published originally because it was part of an unpublished review (as you mentioned in https://github.com/microsoft/LightGBM/pull/6641#issuecomment-2416457468).

I've pushed commits reverting this, for the reasons explained in https://github.com/microsoft/LightGBM/pull/6641#discussion_r1804009897

**Review by jameslamb — COMMENTED** [review]
I've pushed commits to finish this up.

Now that I have commits on here, my approval shouldn't count toward a merge. Hopefully another reviewer will be able to look soon and we can merge this.

After that, I'll put up a PR to start running `shellcheck` in LightGBM's CI. Thanks for your help with this project, @vnherdeiro !

**Review by StrikerRUS — APPROVED** [review]
LGTM!
